modules/e2eping/e2eping.js

```diff
@@ -0,0 +1,333 @@
+    /**
+     * @param {JitsiConference} conference
+     * @param {Object} options
+     */
+    constructor(conference, options) {
+        this.conference = conference;
+
+        // The interval at which pings will be sent (<= 0 disables sending).
+        this.pingIntervalMs = 10000;
+
+        // The interval at which analytics events will be sent.
+        this.analyticsIntervalMs = 60000;
+
+        // Maps a participant ID to its ParticipantState
+        this.participants = {};
+
+        // Whether the WebRTC channel has been opened or not.
+        this.dataChannelOpened = false;
+
+        if (options && options.e2eping) {
+            if (typeof options.e2eping.pingInterval === 'number') {
+                this.pingIntervalMs = options.e2eping.pingInterval;
+            }
+            if (typeof options.e2eping.analyticsInterval === 'number') {
+                this.analyticsIntervalMs = options.e2eping.analyticsInterval;
+            }
+
+            // We want to report at most once a ping interval.
+            if (this.analyticsIntervalMs > 0 && this.analyticsIntervalMs
+                < this.pingIntervalMs) {
+                this.analyticsIntervalMs = this.pingIntervalMs;
+            }
+        }
+        logger.info(
+            `Initializing e2e ping; pingInterval=${
+                this.pingIntervalMs}, analyticsInterval=${
+                this.analyticsIntervalMs}.`);
+
+        // Only subscribe to user join/leave events if sending pings is enabled.
+        if (this.pingIntervalMs > 0) {
+            conference.on(
+                JitsiConferenceEvents.USER_JOINED,
+                (id, participant) => this.participantJoined(participant));
+            conference.on(
+                JitsiConferenceEvents.USER_LEFT,
+                (id, participant) => this.participantLeft(participant));
+        }
+
+        // Listen to E2E PING requests and responses from other participants
+        // in the conference.
+        conference.on(
+            JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            (participant, payload) => {
+                if (payload.type === E2E_PING_REQUEST) {
+                    this.handleRequest(participant, payload);
+                } else if (payload.type === E2E_PING_RESPONSE) {
+                    this.handleResponse(participant, payload);
+                }
+            });
+
+        // We don't want to wait the whole interval before sending the first
+        // request, but we can't send it immediately after the participant joins
+        // either, because our data channel might not have initialized.
+        // So once the data channel initializes, send requests to everyone.
+        // Wait an additional 200ms to give a chance to the remote side (if it
+        // also just connected as is the case for the first 2 participants in a
+        // conference) to open its data channel.
+        conference.on(
+            JitsiConferenceEvents.DATA_CHANNEL_OPENED,
```

Comment:
should this be relying on the more generic 'bridge channel' (which can be using websocket or data channel) instead of datachannel directly?

Reply:
It does, it's just that the event has the old data channel specific name.